lib/perl/Genome/Model/Command/Admin/PurgeSoftwareResultsFromAnalysisProject.pm

```diff
@@ -0,0 +1,171 @@
+        INNER JOIN config.profile_item cpi ON cpi.id = bridge.profile_item_id
+        WHERE cpi.status = 'active'
+        AND candidates.result_id = sr.id
+    ) AND NOT EXISTS (
+        SELECT sr.id
+        FROM result.software_result sr
+        INNER JOIN result."user" sru ON sr.id = sru.software_result_id
+        WHERE sru.user_class_name = 'Genome::Sys::User'
+        AND candidates.result_id = sr.id
+    )
+);
+
+sub execute {
+    my $self = shift;
+
+    foreach my $anp ( $self->analysis_projects ) {
+        $self->purge_one_analysis_project($anp);
+    }
+}
+
+sub analysis_project_is_old_enough_to_purge {
+    my($self, $anp) = @_;
+
+    my $strp = DateTime::Format::Strptime->new(
+        pattern   => UR::Context->date_template(),
+    );
+
+    my $now_str = UR::Context->now();
+    my $updated_at_str = $anp->updated_at;
+
+    my $dt1 = $strp->parse_datetime($updated_at_str);
+    my $dt2 = $strp->parse_datetime($now_str);
+
+    my $anp_updated_duration = $dt1->delta_days($dt2);
+
+    my $duration_to_retain = DateTime::Duration->new(
+        days        => $self->days_to_retain,
+    );
+
+    return ( DateTime::Duration->compare( $anp_updated_duration, $duration_to_retain ) == -1 );
+}
+
+
+sub _get_lock_for_analysis_project {
+    my($self, $anp) = @_;
+
+    my $unlocker;
+    if ($ENV{UR_DBI_NO_COMMIT}) {
+        $unlocker = sub {};
+
+    } else {
+        my $lock = Genome::Sys::LockProxy->new(
+            resource => 'genome model admin purge-software-results-from-analysis-project' . $anp->id,
+            scope => 'site',
+        )->lock(max_try => 1);
+
+        die $self->error_message('Cannot get lock for analysis project '.$anp->id) unless $lock;
+
+        $unlocker = sub {
+            $lock->unlock();
+        };
+    }
+    return $unlocker;
+}
+
+sub purge_one_analysis_project {
+    my($self, $anp) = @_;
+
+    if ($anp->is_cle) {
+        die $self->error_message('Failed to process CLE analysis project: '. $anp->id);
+    }
+
+    if ( $self->analysis_project_is_old_enough_to_purge($anp) ) {
+        $self->warning_message('Analysis project \''. $anp->id .'\' was updated at '. $anp->updated_at .' which is less than the '. $self->days_to_retain .' days to retain disabled results');
+        return;
+    }
+
+    my $dbh = Genome::DataSource::GMSchema->get_default_handle();
+    die $self->error_message('Cannot get default database handle') unless $dbh;
+
+    my $unlocker = $self->_get_lock_for_analysis_project($anp);
+
+    my $total_kb_purged = 0;
+    my $software_result_count = 0;
+
+    my $unlock_and_print_report = Scope::Guard->new(sub {
+        $unlocker->();
+        $self->status_message('Removed %d GB from %d software results',
+                                int($total_kb_purged / KB_IN_ONE_GB),
+                                $software_result_count);
+    });
+
+    my $sth = $dbh->prepare($sql);
+    die $self->error_message("preparing SQL failed : $DBI::errstr") unless $sth;
+
```

Comment:
Maybe prepare the SQL statement handle outside the purging AnP method? And then reuse the $sth to execute?